relay-metrics/src/router.rs

```diff
@@ -0,0 +1,240 @@
+
+                    else => break,
+                }
+            }
+            relay_log::info!("metrics router stopped");
+        });
+    }
+}
+
+/// Generalization of [`InsertMetrics`] and [`MergeBuckets`].
+///
+/// Used to handle these messages generically.
+trait Insert {
+    type Item: MetricsContainer;
+    fn into_parts(self) -> (ProjectKey, Vec<Self::Item>);
+    fn from_parts(project_key: ProjectKey, items: Vec<Self::Item>) -> Self;
+}
+
+impl Insert for InsertMetrics {
+    type Item = Metric;
+
+    fn into_parts(self) -> (ProjectKey, Vec<Self::Item>) {
+        let InsertMetrics {
+            project_key,
+            metrics,
+        } = self;
+        (project_key, metrics)
+    }
+
+    fn from_parts(project_key: ProjectKey, items: Vec<Self::Item>) -> Self {
+        Self {
+            project_key,
+            metrics: items,
+        }
+    }
+}
+
+impl Insert for MergeBuckets {
+    type Item = Bucket;
+
+    fn into_parts(self) -> (ProjectKey, Vec<Self::Item>) {
+        let MergeBuckets {
+            project_key,
+            buckets,
+        } = self;
+        (project_key, buckets)
+    }
+
+    fn from_parts(project_key: ProjectKey, items: Vec<Self::Item>) -> Self {
+        Self {
+            project_key,
+            buckets: items,
+        }
+    }
+}
+
+/// Helper struct that holds the [`Addr`]s of started aggregators.
+struct StartedRouter {
+    default_aggregator: Addr<Aggregator>,
+    secondary_aggregators: BTreeMap<MetricNamespace, Addr<Aggregator>>,
+}
+
+impl StartedRouter {
+    fn start(router: RouterService) -> Self {
+        Self {
+            default_aggregator: router.default_aggregator.start(),
+            secondary_aggregators: router
+                .secondary_aggregators
+                .into_iter()
+                .map(|(key, service)| (key, service.start()))
+                .collect(),
+        }
+    }
+
+    fn handle_message(&mut self, msg: Aggregator) {
+        match msg {
+            Aggregator::AcceptsMetrics(_, sender) => {
+                let requests: Vec<_> = Some(self.default_aggregator.send(AcceptsMetrics))
+                    .into_iter()
+                    .chain(
+                        self.secondary_aggregators
+                            .values_mut()
+                            .map(|agg| agg.send(AcceptsMetrics)),
+                    )
+                    .collect();
+                tokio::spawn(async {
+                    let mut accepts = true;
+                    for req in requests {
+                        accepts &= req.await.unwrap_or_default();
+                    }
+                    sender.send(accepts);
+                });
+            }
+            Aggregator::InsertMetrics(msg) => self.handle_metrics(msg),
+            Aggregator::MergeBuckets(msg) => self.handle_metrics(msg),
+            #[cfg(test)]
+            Aggregator::BucketCountInquiry(_, _sender) => (), // not supported
```

Comment:
I suppose this is fine because the tests will simply use the aggregator service directly?

Reply:
Yeah it's not ideal, but acceptable IMO because this message is only used in one test.